### PR TITLE
Refactor info to support multiple regions

### PR DESCRIPTION
**Pull request by thedadams**

### CODE REVIEW

File: pkg/cli/info.go [hunk 1]
```diff
@@ -52,11 +57,27 @@ func (s *Info) Run(cmd *cobra.Command, args []string) error {
...
+		if s.Verbose {
+			// If the user passed the verbose flag, then give them the entire info response
+			projectInfo[subInfo.Name] = subInfo.Regions
+			continue
+		}
+
+		// Remove unset fields in the config and userConfig
+		regionInfo := make(map[string]any, len(subInfo.Regions))
+		for region, spec := range subInfo.Regions {
+			specMap, err := removeUnsetFields(spec, "config", "userConfig")
```
Comment (njhale) [inline pkg/cli/info.go:73]: It feels like removing unset fields is a general formatting problem that can be solved at the `table.Writer` level; either by exposing some first class feature of `table.Writer` or by creating a new delegating writer type that proxies write calls, dropping the fields before passing the object along.

File: pkg/cli/info.go [hunk 2]
```diff
@@ -72,3 +93,30 @@ func (s *Info) Run(cmd *cobra.Command, args []string) error {
...
+					delete(cfg, entry.Key)
+				}
+			}
+			if len(cfg) == 0 {
+				delete(specMap, key)
+			}
+		}
+	}
+
+	return specMap, nil
```
Comment (njhale) [inline pkg/cli/info.go:121]: I _think_ there's a more concise way to do this involving [gojq](https://pkg.go.dev/github.com/itchyny/gojq#Parse) and a query like `del(.. | select(. == null or . == ""))`.

**Review by njhale — COMMENTED** [review]
Looking great to me so far!

Just have a few questions and suggestions

File: integration/helper/secrets.go [hunk 3]
```diff
@@ -18,8 +18,10 @@ func EncryptData(t *testing.T, client client.Client, keys []string, data string)
 			return ""
 		}
 		for _, info := range fullInfo {
-			for _, key := range info.Spec.PublicKeys {
-				pubKeys = append(pubKeys, key.KeyID)
+			for _, region := range info.Regions {
+				for _, key := range region.PublicKeys {
+					pubKeys = append(pubKeys, key.KeyID)
+				}
 			}
```
[the remark this answers is not in the document]
Comment (thedadams) [reply integration/helper/secrets.go:25]: I will take a look. I was just refactoring to make it work after the changes.

File: pkg/apis/api.acorn.io/v1/types.go [hunk 4]
```diff
@@ -352,17 +353,16 @@ type BuilderPortOptions struct {
...
-	Config                 Config            `json:"config"`
-	UserConfig             Config            `json:"userConfig"`
-	LetsEncryptCertificate string            `json:"letsEncryptCertificate,omitempty"`
-	ExtraData              map[string]string `json:"extraData,omitempty"`
+	Version                string          `json:"version"`
+	Tag                    string          `json:"tag"`
+	GitCommit              string          `json:"gitCommit"`
+	Dirty                  bool            `json:"dirty"`
+	ControllerImage        string          `json:"controllerImage"`
+	APIServerImage         string          `json:"apiServerImage,omitempty"`
```
[the remark this answers is not in the document]
Comment (thedadams) [reply pkg/apis/api.acorn.io/v1/types.go:361]: There is none now, but it was decided that it was very likely to happen in the near future.

File: pkg/apis/api.acorn.io/v1/types.go [hunk 5]
```diff
@@ -341,7 +341,8 @@ type Info struct {
 	metav1.TypeMeta   `json:",inline"`
 	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
 
-	Spec InfoSpec `json:"spec,omitempty"`
+	Regions   map[string]InfoSpec `json:"regions,omitempty"`
+	ExtraData map[string]string   `json:"extraData,omitempty"`
```
[the remark this answers is not in the document]
Comment (thedadams) [reply pkg/apis/api.acorn.io/v1/types.go:345]: The ExtraData is only used to store Hub information right now, which is at the top-level.

An ExtraData field can be added at the region level later, if needed.

File: pkg/cli/info.go [hunk 6]
```diff
@@ -1,12 +1,16 @@
 package cli
 
 import (
-	apiv1 "github.com/acorn-io/acorn/pkg/apis/api.acorn.io/v1"
+	"encoding/json"
+	"reflect"
+
+	v1 "github.com/acorn-io/acorn/pkg/apis/api.acorn.io/v1"
```
[the remark this answers is not in the document]
Comment (thedadams) [reply pkg/cli/info.go:7]: This is required. Go treats `/v*` as package versions. In order to reference this package as `v1`, this name is needed.

File: pkg/cli/info.go (hunk 1 above)
Comment (thedadams) [reply pkg/cli/info.go:73]: The issue is that the table writer uses json/yaml marshaling. That means that in order to get the table writer to do this, we would have to use `omitempty`. However, we cannot do that with the config fields (see the comment on the Config type).

This implantation is essentially doing the "proxy" idea. Do you think it is worth it to try to make this more general right now?

File: pkg/cli/info.go (hunk 2 above)
Comment (thedadams) [reply pkg/cli/info.go:121]: I'm not familiar with gojq, so your suggestion doesn't seem as understandable or concise to me. I would also be curious whether using gojq would be less performant than using built-in Go functionality.

Do you think it would be worth it to explore the performance trade offs?

File: pkg/apis/api.acorn.io/v1/types.go [hunk 7]
```diff
@@ -341,7 +341,8 @@ type Info struct {
 	metav1.TypeMeta   `json:",inline"`
 	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
 
-	Spec InfoSpec `json:"spec,omitempty"`
+	Regions   map[string]InfoSpec `json:"regions,omitempty"`
```
[the remark this answers is not in the document]
Comment (thedadams) [reply pkg/apis/api.acorn.io/v1/types.go:344]: Could you elaborate on this? Just want to understand.

My opinion is that adding this named typed makes the code less readable because I don't know what the type InfoSpecs is. I would have to look at it to see that it is just a map in disguise.

That is just my opinion though. It could be wrong and am interested to hear more about yours.

**Review by njhale — APPROVED** [review]
This PR looks functionally correct to me, and I think it can go in as is -- the suggestions I left in my last review are things that, if we decided we wanted, can always be addressed later.

File: docs/docs/100-reference/01-command-line/acorn_info.md [hunk 8]
```diff
@@ -14,6 +14,7 @@ acorn info [flags]
 ```
   -h, --help            help for info
   -o, --output string   Output format (json, yaml, {{gotemplate}}) (default "yaml")
+      --verbose         Print entire Acorn config including values not configured
```
[the remark this answers is not in the document]
Comment (njhale) [reply docs/docs/100-reference/01-command-line/acorn_info.md:17]: got it! thanks for confirming

File: pkg/apis/api.acorn.io/v1/types.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (njhale) [reply pkg/apis/api.acorn.io/v1/types.go:344]: > Could you elaborate on this? Just want to understand.

This is mostly subjective, but in my opinion the extra types:
- serve as a way to better codify the semantics of the map key
- have slightly more utility since they can receive helper methods
- are a bit less clunky to use; i.e. the maps type literal doesn't proliferate
  ```go
    func foo(specs InfoSpecs) { //...
    // vs.
    func foo(specs map[string]InfoSpec) { //...

    specs := make(InfoSpecs, 10)
    // vs.
    specs := make(map[string]InfoSpec, 10)
  ```

> My opinion is that adding this named typed makes the code less readable because I don't know what the type InfoSpecs is. I would have to look at it to see that it is just a map in disguise.

Yeah, that's a great point. Now that you mentioned it, I can see how the extra layers of indirection could be misleading and I'd rather not introduce a footgun for my future self.

[the remark this answers is not in the document]
Comment (thedadams) [reply pkg/apis/api.acorn.io/v1/types.go:344]: Thank you for including your thoughts here. I think they are definitely valid reasons for including a type.

File: pkg/cli/info.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (njhale) [reply pkg/cli/info.go:7]: > Go treats /v* as package versions

So IIUC, for a given package import, it's actually the top-level _module path_ that determines the major version of the module (and package) resolved. 

Barring any nested Modules, the _module path_ here is `github.com/acorn-io/acorn`, so that `v1` at the very end of the import statement shouldn't affect the version of the package that gets resolved.

but maybe I'm missing something

File: pkg/cli/info.go (hunk 1 above)
Comment (njhale) [reply pkg/cli/info.go:73]: > Do you think it is worth it to try to make this more general right now?

No, I don't think so. I imagine we'd only pursue this if/when we need similar behavior in other commands.

File: pkg/cli/info.go (hunk 2 above)
Comment (njhale) [reply pkg/cli/info.go:121]: > I'm not familiar with gojq, so your suggestion doesn't seem as understandable or concise to me.

TLDR; Please ignore this suggestion! Sorry!

I used it to solve a similar problem in the past, but it turned out the example I was thinking of wasn't as clean as I remembered. Meant to close this out earlier.

File: pkg/cli/info.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (thedadams) [reply pkg/cli/info.go:7]: No, you are absolutely right here. I just tested it and it seems fine to remove this alias.

There was an instance previously where such an alias was necessary. Not sure what the difference is. I will remove this one.

**Review by g-linville — APPROVED** [review]

**Review by tylerslaton — APPROVED** [review]
